### PR TITLE
fix(lark): use constant_time_eq for verification_token comparison

### PR DESCRIPTION
**Pull request by tzy-17**
## Summary

- **Base branch:** `master`
- **What changed and why:** Lark URL verification challenge at `lark.rs` compared `verification_token` with `==`, which short-circuits on first mismatched byte and is vulnerable to timing attacks. Replaced with `constant_time_eq()` from `zeroclaw_config::pairing` — the same function already used for node auth tokens (`nodes.rs:264`), webhook secrets (`gateway/lib.rs:2723`), WhatsApp verify tokens (`gateway/lib.rs:3039`), and OTP pairing codes (`pairing.rs:139`). Additionally made the verification fail-closed: missing, null, non-string, or empty incoming tokens are now rejected instead of accepted. Extracted the logic into a standalone `verify_challenge_token()` function for testability, and added handler-level regression tests covering all edge cases.
- **Scope boundary:** Only changed the URL verification challenge handler; no other Lark endpoints touched.
- **Blast radius:** Lark and Feishu channel URL verification endpoints only. The token is still read from the selected Lark instance config; no config schema changes. **Breaking behavior change:** deployments that previously accepted URL verification challenges without a token in the active alias will now reject those challenges. Operators must ensure `verification_token` is configured under the selected instance, such as `[channels.lark.default]`; Feishu uses the same Lark instance with `use_feishu = true`.
- **Linked issue(s):** None filed yet — security fix discovered during code review.
- **Labels:** `bug`, `ci`, `channel`, `trusted contributor`, `channel:lark`, `domain:security`, `risk:high`, `size:M`

### How you can test (when useful)

- **Reviewer testing requested?** Yes
- **Interface(s) exercised:** `channel:lark` (webhook callback, including Feishu when the selected Lark instance has `use_feishu = true`)
- **Setup / preconditions:** A Lark or Feishu app configured in webhook mode with `verification_token` set under the selected alias, such as `[channels.lark.default]`.
- **Steps to run:** Send a URL verification challenge to the `/lark` endpoint with various token payloads:
  1. `{"challenge": "abc", "token": "<correct_token>"}` → should return 200 with `{"challenge": "abc"}`
  2. `{"challenge": "abc", "token": "wrong_token"}` → should return 403
  3. `{"challenge": "abc"}` (no token field) → should return 403
  4. `{"challenge": "abc", "token": null}` → should return 403
  5. `{"challenge": "abc", "token": 12345}` → should return 403
  6. `{"challenge": "abc", "token": ""}` → should return 403
- **Expected on this branch (after):** Only case 1 returns 200; all others return 403.
- **Prior behavior on `master` (before):** Cases 3 and 4 were accepted (200); case 6 with an empty configured token was also accepted. The `==` comparison was also vulnerable to timing attacks.

### How I tested

- **CI checks relied on and why they cover this change:** Fresh required CI is green at exact head `89ef000e030abb085f4342dd501fa5b6f531ea88`. The required `Test` job explicitly runs `cargo nextest run --locked -p zeroclaw-channels --no-default-features --features channel-lark --lib lark_url_verification` and reports all 9 focused tests passing, so it executes the feature-gated handler regressions; `CI Required Gate` is also green on that head.
- **Known CI coverage gap, if any:** No known CI gap for the feature-gated Lark URL-verification regressions. No live HTTP listener smoke was run; the handler-level tests cover the production status/body decision without binding a listener.
- **Commands run and tail output:** `zc-cargo fmt --all -- --check` passed. Exact-head review validation recorded `cargo +1.96.1 nextest run --locked -p zeroclaw-channels --no-default-features --features channel-lark --lib lark_url_verification` with all 9 focused tests passing, plus feature-enabled Clippy and the full Lark library test set passing. An earlier local attempt did not reach tests because the SSD ran out of space while compiling unrelated Matrix crates; it is not treated as passing evidence.
- **Beyond CI, what did you manually verify?** Reviewed the fail-closed `verify_challenge_token` behavior, the production handler call, and the handler-level cases for correct, wrong, missing, null, non-string, and empty tokens.
- **If any command was intentionally skipped, why:** No live HTTP callback smoke was run because the handler tests exercise the production decision without binding a listener. No broad local Cargo retry was run after the earlier SSD-space failure; fresh hosted CI provides the workspace execution proof.

## Security & Privacy Impact (required)

- New permissions, capabilities, or file system access scope? No.
- New external network calls? No.
- Secrets / tokens / credentials handling changed? Yes — existing `verification_token` is now compared with constant-time comparison instead of `==`, and the verification is fail-closed (missing/null/non-string/empty tokens are rejected). Previously, missing or null tokens were accepted, and the comparison was vulnerable to timing side-channel leakage.
- PII, real identities, or personal data in diff, tests, fixtures, or docs? No.
- Prompt injection or untrusted model-visible text introduced/changed? No.
- If any `Yes`, describe the risk and mitigation: The token comparison change eliminates a timing side-channel. The fail-closed behavior change means operators must configure `verification_token` for the selected Lark alias before webhook mode will pass URL verification; this is the correct security posture per FND-006 §4.5.

## Compatibility (required)

- Backward compatible? No — deployments without `verification_token` configured for the active Lark alias will have URL verification challenges rejected instead of accepted.
- Config / env / CLI surface changed? No.
- Rust/MSRV/toolchain floor changed? No.
- If backward compatibility is `No` or either surface/floor question is `Yes`: exact upgrade steps for existing users: Set `verification_token` under the selected Lark instance, for example `[channels.lark.default]`, or replace `default` with the configured alias. Feishu webhook mode uses the same Lark instance with `use_feishu = true`.

## Rollback (required for medium/high-risk PRs)

- **Fast rollback command/path:** Revert the landed #9110 squash-merge commit, then rerun the focused Lark URL-verification tests and required CI.
- **Feature flags or config toggles:** None.
- **Observable failure symptoms:** A deployment whose active Lark alias lacks `verification_token` will receive `403 Forbidden` from `/lark` URL verification. Confirm the token is present under the selected alias before considering rollback. Rolling back restores the fail-open behavior and the timing-sensitive comparison, so use it only as an emergency measure.

### CODE REVIEW

**Review by Audacity88 — APPROVED** [review]
@tzy-17 I checked the current head, the one-file diff, the public Lark URL-verification contract, the existing constant-time helper and its tests, and the green required CI. There are no prior reviews, review comments, or linked issues. I did not run additional local Cargo validation.

### 🟢 What looks good — Reuses the established constant-time primitive

The production change is small and coherent: it replaces the ordinary equality check with the repository's existing `constant_time_eq` helper without introducing another token-comparison implementation.

### 🟡 Warning — URL verification still accepts missing, malformed, or empty tokens

The surrounding `is_none_or` behavior predates this PR, so this is not a blocker for the constant-time substitution. It does mean a challenge with an absent, `null`, or non-string `token` is still accepted without comparison, while an empty configured token accepts an explicitly empty incoming token. The documented URL-verification request includes the application verification token as a string. Please route this existing fail-open behavior through the appropriate security follow-up rather than treating this PR as complete hardening of the callback boundary.

### 🔵 Suggestion — Cover the Lark decision point

The new test calls `constant_time_eq` directly, duplicating coverage already present with the helper. It would still pass if the Lark handler reverted to `==`. A handler-level regression around the Lark token decision would provide more useful protection, particularly alongside the fail-closed follow-up above.

The scoped constant-time comparison change itself looks good to merge after the branch is refreshed from current `master` and required CI reruns.

**Review by IftekharUddin — CHANGES_REQUESTED** [review]
@tzy-17 I reviewed d235bac62ea71d067ab05751be5fe7df6a8efc76, including the complete one-file diff, the callback handler and configuration path, the existing constant-time helper and its tests, the prior maintainer approval, the current PR template, and the exact-head required CI. There are no inline threads or linked issues. The hosted required gate is green; I did not duplicate its Cargo run locally.

### 🟢 What looks good — Reuses the established comparison primitive

Using `zeroclaw_config::pairing::constant_time_eq` keeps token comparison on the repository's existing, reviewed primitive instead of adding another implementation at the channel boundary.

### 🔴 Blocking — The modified verification path still fails open

The new expression retains `is_none_or`, so a challenge with an absent, `null`, or non-string `token` bypasses comparison and is accepted. `LarkConfig::verification_token` is optional and becomes an empty string in the channel constructor, which also allows an explicitly empty incoming token. This is the exact trust-boundary expression changed by the PR, and FND-006 §4.5 requires externally supplied data to be validated and security failures to fail loudly. Please make the challenge require a string token that matches a non-empty configured verification token, and cover missing, malformed, empty, wrong, and correct values through the handler decision point.

The added test currently calls `constant_time_eq` directly, duplicating the helper's existing tests; it would stay green if this Lark handler reverted to `==` or continued accepting a missing token. A handler-level regression is needed to prove the security behavior being changed.

### 🔴 Blocking — Public review evidence does not satisfy the current high-risk template

The body still uses the retired `Validation Evidence` shape and omits the current required testing prompts, the complete security/privacy checklist, the current compatibility questions, and the high-risk rollback fields. Please update it against `.github/pull_request_template.md`, including the exact hosted checks relied on, the known coverage gap above, whether reviewer testing is useful, and observable rollback symptoms. This is substantive review evidence for a `risk:high` security change, not a heading-only cleanup.

The existing approval's warning about fail-open token handling remains unresolved, and no security follow-up issue was linked. I cannot reconcile that concern as settled on this head.

File: crates/zeroclaw-channels/src/lark.rs [hunk 1]
```diff
@@ -3713,7 +3714,7 @@ impl LarkChannel {
                 let token_ok = payload
                     .get("token")
                     .and_then(|t| t.as_str())
-                    .is_none_or(|t| t == state.verification_token);
+                    .is_none_or(|t| constant_time_eq(t, &state.verification_token));
```
Comment (IftekharUddin) [inline crates/zeroclaw-channels/src/lark.rs]: 🔴 **Blocking:** This exact expression still fails open: an absent, `null`, or non-string token makes `is_none_or` return true, and the optional configured token is converted to an empty string, which accepts an explicitly empty token. Because this is an external verification boundary, require a non-empty configured token and an incoming string that matches it; missing, malformed, empty, and wrong values should all be rejected.

File: crates/zeroclaw-channels/src/lark.rs [hunk 2]
```diff
@@ -4196,6 +4197,14 @@ mod tests {
         )
     }
 
+    #[test]
+    fn lark_verification_token_comparison_is_constant_time() {
```
Comment (IftekharUddin) [inline crates/zeroclaw-channels/src/lark.rs]: 🔴 **Blocking:** This test exercises the shared helper directly, which is already covered in `zeroclaw-config`; it would pass if the Lark handler reverted to `==` or continued accepting missing tokens. Please exercise the handler's observable accept/reject decision for correct, wrong, missing, malformed, and empty tokens so the regression proves the security behavior this PR changes.

**Comment by Audacity88** [comment]
This PR is still waiting on the requested fail-closed verification-token fix. A missing, null, or non-string token must be rejected, with regressions covering those cases alongside the valid constant-time comparison path.

Please update the branch or reply by **August 8, 2026**. Without an update, we will close this PR for backlog hygiene. The Lark verification hardening can return in a refreshed PR.

**Comment by tzy-17** [comment]
> This PR is still waiting on the requested fail-closed verification-token fix. A missing, null, or non-string token must be rejected, with regressions covering those cases alongside the valid constant-time comparison path.
> 
> Please update the branch or reply by **August 8, 2026**. Without an update, we will close this PR for backlog hygiene. The Lark verification hardening can return in a refreshed PR.

Thanks for the reminder! I’ll address the fail-closed verification-token fix, including proper rejection for missing, null, or non-string tokens, and add regression tests covering those cases along with the constant-time comparison path. I’ll have the branch updated by August 8, 2026.

**Comment by Audacity88** [comment]
@tzy-17, the requested handler-boundary coverage remains unanswered after the July 29 update.

Marking this as a stale candidate. Please add a regression through the actual Lark URL-verification decision point covering missing, malformed, empty, incorrect, and correct tokens, refresh the validation record, or reply with your plan by **August 20, 2026**.

The fail-closed token fix remains useful. A stale closure would retire this branch, not reject that security correction; the work could return in a refreshed PR.

**Comment by Audacity88** [comment]
## Status Update

The Lark verification repair is now a clean, reviewable three-commit change on `3deefdd2`. The two contributor commits preserve constant-time token comparison and fail-closed challenge handling. A separate callback-hardening commit adds direct handler-boundary coverage.

The current diff is confined to `crates/zeroclaw-channels/src/lark.rs`. I also corrected an earlier takeover-history mistake that made GitHub show unrelated upstream commits in this PR. The replacement branch is linear and preserves the contributor work separately from the maintainer hardening.

Fresh CI is running on `3deefdd2`. The existing changes-requested review remains in place. After CI finishes, this needs independent re-review; it is no longer waiting on contributor action.

**Comment by Audacity88** [comment]
The feature-gated coverage gap is fixed at `c18211e745`.

The required `Test` job now runs the Lark URL-verification handler regressions with `--no-default-features --features channel-lark`; that step passed, and `CI Required Gate` is green. I also corrected the PR body to name this lane and the local SSD-space limitation accurately.

The remaining `CHANGES_REQUESTED` state is from the earlier head. This repair now needs independent maintainer re-review on the current green head.

**Review by IftekharUddin — APPROVED** [review]
@tzy-17 I re-reviewed `c18211e745b811cce584f71dc62b4eeccbed91d8`, including the complete two-file diff, the Lark webhook configuration and handler path, both earlier inline threads, the current PR template and labels, FND-004, FND-006, and the fresh hosted checks. The required gate is green, and the required `Test` job log records all nine focused `lark_url_verification` tests passing. I did not check out or execute the contributor branch locally.

### ✅ Resolved — Challenge verification now fails closed

The earlier verification-boundary blocker is resolved. `verify_challenge_token` now requires both the configured token and incoming string token to be non-empty before using the repository's established constant-time comparator. The production handler returns `403` for missing, null, non-string, empty, and wrong tokens, while a valid token receives the challenge response.

### ✅ Resolved — Regressions exercise the production decision point

The earlier helper-only test blocker is resolved. The new handler-level cases call `handle_lark_http_event` and assert the observable status and body for invalid tokens, an empty configured token, malformed challenges, and the valid challenge path. These tests would fail if the handler returned to the old fail-open expression.

### ✅ Resolved — High-risk review evidence is current

The PR body now answers the current testing, security/privacy, compatibility, and rollback prompts, explicitly identifies the operator-facing behavior change, and gives the required configuration step. The CI change places the focused feature-enabled command inside the required `Test` job, and the aggregate gate depends on that job.

### 🟢 What looks good — Scope stays at the Lark verification boundary

The extraction preserves the non-challenge event dispatch path while making the externally supplied URL-verification challenge fail closed, consistent with FND-006 §4.5. It reuses the shared comparison primitive and introduces no new dependency, permission, configuration surface, or network call.

### 🟡 Warning — A second independent exact-head approval is still required

This remains a `risk:high` security-boundary change. My approval resolves my prior changes-requested state, but it does not replace the repository's second independent Core Team approval requirement for the current head.

I found no new blocking issue on this head. Approved.

**Review by Audacity88 — APPROVED** [review]
@tzy-17,

I reviewed the current exact head, 89ef000e030abb085f4342dd501fa5b6f531ea88, including the complete Lark verification path, the contributor-authored changes, my later repair commits, the prior human and automated feedback, and the current hosted checks.

For authorship clarity: you authored the original constant-time comparison and the fail-closed verifier. I later authored and pushed the handler-boundary hardening, the direct status/body regression tests, and the feature-enabled Lark CI lane. Those repairs preserve the purpose of your contribution—closing the Lark verification-token boundary—and add no new scope or risk family.

### ✅ Resolved — Verification now fails closed at the production boundary

The current handler rejects missing, null, non-string, empty, and incorrect tokens, while accepting a correct non-empty token through the established constant-time comparator. The handler-level cases exercise the observable status and response body, resolving @IftekharUddin's earlier boundary and regression findings.

### ✅ Resolved — Exact-head validation evidence is now accurate

The exact-head automated review found the substantive security behavior, tests, and CI integration resolved. Its sole remaining finding was that the PR body still cited an older validation SHA. I corrected that public record to 89ef000e030abb085f4342dd501fa5b6f531ea88 and verified the readback. The required gate is green on this head, and the required Test job ran all nine focused Lark URL-verification tests successfully.

### 🟢 What looks good — Narrow repair with explicit operator impact

The change stays within the Lark and Feishu URL-verification boundary, reuses the shared comparison primitive, and clearly documents the fail-closed compatibility impact and configuration requirement.

Because I contributed the later repairs, this is my accountable Core Team approval after independently reviewing the repaired aggregate head and reconciling the exact-head bot finding. The bot supplies technical evidence, not an approval. @IftekharUddin's independent approval on this head supplies the second Core Team approval required for this `risk:high` change.

Approved.
